lightning/src/ln/channelmanager.rs

```diff
@@ -3819,6 +3823,15 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 									 HTLCSource::PreviousHopData(htlc.prev_hop), &payment_hash,
 									 HTLCFailReason::Reason { failure_code: 0x4000|15, data: htlc_msat_height_data });
 				} else {
+					if purpose.is_none() {
+						purpose = Some(match htlc.onion_payload {
+							OnionPayload::Spontaneous(preimage) => events::PaymentPurpose::SpontaneousPayment(preimage),
+							OnionPayload::Invoice(hop_data) => events::PaymentPurpose::InvoicePayment {
+								payment_secret: hop_data.payment_secret,
+								payment_preimage: Some(payment_preimage),
+							}
+						});
+					}
 					match self.claim_funds_from_hop(channel_state.as_mut().unwrap(), htlc.prev_hop, payment_preimage) {
 						ClaimFundsFromHop::MonitorUpdateFail(pk, err, _) => {
 							if let msgs::ErrorAction::IgnoreError = err.err.action {
```

Comment:
Yea, I mean there's no good answer here (and this PR doesn't change that). TBH I think we need to rethink the whole perm-failure logic a bit. That's kinda the intent behind  #1106, but ultimately the "what does perm-failure *mean*" topic isn't super well-defined currently. Lets leave this for a followup because there's nothing changed here, but, does perm-failure mean "we updated the in-memory copy and nothing else" or "we failed to update anything" needs addressing.